File: src/main/java/org/lanternpowered/server/advancement/LanternAdvancementBuilder.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.advancement;

import static com.google.common.base.Preconditions.checkNotNull;

import org.lanternpowered.server.catalog.AbstractNamedCatalogBuilder;
import org.spongepowered.api.ResourceKey;
import org.spongepowered.api.advancement.Advancement;
import org.spongepowered.api.advancement.DisplayInfo;
import org.spongepowered.api.advancement.criteria.AdvancementCriterion;
import org.spongepowered.api.text.translation.FixedTranslation;
import org.spongepowered.api.text.translation.Translation;

import org.checkerframework.checker.nullness.qual.Nullable;

@SuppressWarnings({"NullableProblems", "ConstantConditions"})
public class LanternAdvancementBuilder extends AbstractNamedCatalogBuilder<Advancement, Advancement.Builder> implements Advancement.Builder {

    @Nullable private Advancement parent;
    private AdvancementCriterion criterion;
    @Nullable private DisplayInfo displayInfo;

    public LanternAdvancementBuilder() {
        reset();
    }

    @Override
    protected Translation getFinalName() {
        if (getName() != null) {
            return getName();
        } else if (this.displayInfo != null) {
            return TextTranslation.of(this.displayInfo.getTitle());
        }
        return new FixedTranslation(getKey().getValue());
    }

    @Override
    public Advancement.Builder parent(@Nullable Advancement parent) {
        this.parent = parent;
        return this;
    }

    @Override
    public Advancement.Builder criterion(AdvancementCriterion criterion) {
        checkNotNull(criterion, "criterion");
        this.criterion = criterion;
        return this;
    }

    @Override
    public Advancement.Builder displayInfo(@Nullable DisplayInfo displayInfo) {
        this.displayInfo = displayInfo;
        return this;
    }

    @Override
    public Advancement.Builder reset() {
        this.criterion = AdvancementCriterion.empty();
        this.displayInfo = null;
        this.parent = null;
        return super.reset();
    }

    @Override
    protected Advancement build(ResourceKey key, Translation name) {
        return new LanternAdvancement(key, name, this.parent, this.displayInfo, this.criterion);
    }
}

File: src/main/kotlin/org/lanternpowered/server/advancement/LanternAdvancementBuilder.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.advancement

import org.lanternpowered.api.key.NamespacedKey
import org.lanternpowered.api.text.toPlain
import org.lanternpowered.server.catalog.AbstractNamedCatalogBuilder
import org.spongepowered.api.advancement.Advancement
import org.spongepowered.api.advancement.DisplayInfo
import org.spongepowered.api.advancement.criteria.AdvancementCriterion

class LanternAdvancementBuilder : AbstractNamedCatalogBuilder<Advancement, Advancement.Builder>(), Advancement.Builder {

    private var parent: Advancement? = null
    private var criterion: AdvancementCriterion = AdvancementCriterion.empty()
    private var displayInfo: DisplayInfo? = null

    override fun getFinalName(key: NamespacedKey): String {
        val name = this.name
        if (name != null)
            return name
        val displayInfo = this.displayInfo
        if (displayInfo != null)
            return displayInfo.title.toPlain()
        return key.value
    }

    override fun parent(parent: Advancement?): Advancement.Builder =
            this.apply { this.parent = parent }

    override fun criterion(criterion: AdvancementCriterion): Advancement.Builder =
            this.apply { this.criterion = criterion }

    override fun displayInfo(displayInfo: DisplayInfo?): Advancement.Builder =
            this.apply { this.displayInfo = displayInfo }

    override fun reset(): Advancement.Builder {
        this.criterion = AdvancementCriterion.empty()
        this.displayInfo = null
        this.parent = null
        return super.reset()
    }

    override fun build(key: NamespacedKey, name: String): Advancement =
            LanternAdvancement(key, name, this.parent, this.displayInfo, this.criterion)
}
